File: src/Samples/FSharpApp/ViewModels/NodeGraph.fs
namespace FSharpApp.ViewModels

open System
open System.Collections.Generic
open System.Diagnostics
open QuickGraph
open GraphSharp.Controls
open Utopia
open Utopia.ViewModel


type INodeVertex = 
    abstract ID : string
    abstract Node : INode

type NodeEdge(source : INodeVertex, target : INodeVertex) = 
    inherit Edge<INodeVertex>(source, target)
    member this.ID : string = sprintf "{%s} -> {%s}" source.ID target.ID
    override this.ToString() = this.ID

type NodeGraph(ce : ICalculationEngine, vertexConstructor : (INode -> INodeVertex)) as this = 
    inherit BidirectionalGraph<INodeVertex, NodeEdge>()
    let vertices = new Dictionary<string, INodeVertex>()
    
    do 
        for n in ce.Nodes do
            let vertex = vertexConstructor(n)
            vertices.Add(n.ID, vertex)
            this.AddVertex(vertex) |> ignore
        for n in ce.Nodes do
            for d in n.DependentNodes do
                let edge = NodeEdge(vertices.[d.ID], vertices.[n.ID])
                this.AddEdge(edge) |> ignore
    
    member this.CalcEngine = ce
    member this.UpdateNode(id : string) = 
        let vertex = vertices.[id]
        vertex.Node.AsyncCalculate()

type NodeGraphLayout() = 
    inherit GraphLayout<INodeVertex, NodeEdge, NodeGraph>()

type NodeVertexBase(node : INode) as this =
    inherit ViewModelBase()

    do
        node.Changed.Add(fun args -> this.RaisePropertyChanged "Dirty")

    member val Node = node
    member val ID = node.ID
    member val IsInput = node.IsInput

    member this.Dirty with get () = node.Dirty
    override this.ToString() = sprintf "{%s}-{%b}" this.ID this.IsInput
                   
    interface INodeVertex with
        member I.ID = this.ID
        member I.Node = this.Node

File: src/Samples/CSharpApp/ViewModels/NodeGraph.cs
namespace CSharpApp.ViewModels
{
    using System;
    using System.Collections.Generic;
    using GraphSharp.Controls;
    using QuickGraph;
    using Utopia;
    using Utopia.ViewModel;

    public interface INodeVertex
    {
        string ID { get; }
        INode Node { get; }
    }

    public class NodeEdge : Edge<INodeVertex>
    {
        public string ID { get; private set; }

        public NodeEdge(INodeVertex source, INodeVertex target)
            : base(source, target)
        {
            ID = string.Format("{0} -> {1}", source.ID, target.ID);

        }

        public override string ToString()
        {
            return ID;
        }
    }

    public class NodeGraph : BidirectionalGraph<INodeVertex, NodeEdge>
    {
        IDictionary<string, INodeVertex> _vertices;

        public NodeGraph(ICalculationEngine calculationEngine, Func<INode, INodeVertex> vertexConstructor)
            : base()
        {
            _vertices = new Dictionary<string, INodeVertex>();

            foreach (var node in calculationEngine.Nodes)
            {
                var vertex = vertexConstructor(node);
                _vertices.Add(node.ID, vertex);
                AddVertex(vertex);
            }

            foreach (var node in calculationEngine.Nodes)
            {
                foreach (var dependentNode in node.DependentNodes)
                {
                    var edge = new NodeEdge(_vertices[dependentNode.ID], _vertices[node.ID]);
                    AddEdge(edge);
                }
            }
        }

        public void UpdateNode(string id)
        {
            var vertex = _vertices[id];
            vertex.Node.AsyncCalculate();
        }

    }

    public class NodeGraphLayout : GraphLayout<INodeVertex, NodeEdge, NodeGraph>
    {
        public NodeGraphLayout() : base() { }
    }

    public class NodeVertexBase : ViewModelBase, INodeVertex
    {
        public NodeVertexBase(INode node)
            : base()
        {
            Node = node;

            Node.Changed += new ChangedEventHandler((sender, args) => RaisePropertyChanged("Dirty"));
        }

        public INode Node { get; private set; }

        public string ID { get { return Node.ID; } }

        public bool IsInput { get { return Node.IsInput; } }

        public bool Dirty { get { return Node.Dirty; } }

        public override string ToString() { return string.Format("{0}-{1}", ID, IsInput); }
    }
}